src/omv/ports/mimxrt/omv_spi.c

```diff
@@ -145,46 +227,64 @@ int omv_spi_transfer_start(omv_spi_t *spi, omv_spi_transfer_t *xfer) {
 }
 
 int omv_spi_transfer_abort(omv_spi_t *spi) {
-    LPSPI_MasterTransferAbortEDMA(spi->inst, &spi->descr_master_edma);
+    if (spi->dma_flags & (OMV_SPI_DMA_NORMAL | OMV_SPI_DMA_CIRCULAR)) {
+        LPSPI_MasterTransferAbortEDMA(spi->inst, &spi->descr_master_edma);
+    }
+    LPSPI_MasterTransferAbort(spi->inst, &spi->descr_master);
+    LPSPI_MasterInit(spi->inst, &spi->config_backup, BOARD_BOOTCLOCKRUN_LPSPI_CLK_ROOT);
     return 0;
 }
 
 static int omv_spi_dma_init(edma_handle_t *dma_handle, const dma_descr_t *dma_descr) {
-    DMAMUX_SetSource(dma_descr->dma_mux, dma_descr->channel, dma_descr->request);
-    DMAMUX_EnableChannel(dma_descr->dma_mux, dma_descr->channel);
-    EDMA_CreateHandle(dma_handle, dma_descr->dma_inst, dma_descr->channel);
+    int channel = allocate_dma_channel();
+    DMAMUX_SetSource(dma_descr->dma_mux, channel, dma_descr->request);
+    DMAMUX_EnableChannel(dma_descr->dma_mux, channel);
+    EDMA_CreateHandle(dma_handle, dma_descr->dma_inst, channel);
     return 0;
 }
 
 int omv_spi_init(omv_spi_t *spi, omv_spi_config_t *config) {
-    // Configure clocks, pins and DMA MUX.
-    mimxrt_hal_spi_init(config->id, config->nss_enable, config->nss_pol);
+    memset(spi, 0, sizeof(omv_spi_t));
 
     const omv_spi_descr_t *spi_descr = &omv_spi_descr_all[config->id - 1];
     if (spi_descr->inst == NULL) {
         return -1;
     }
 
-    memset(spi, 0, sizeof(omv_spi_t));
     spi->id = config->id;
     spi->inst = spi_descr->inst;
     spi->cs = spi_descr->cs;
-    spi->dma_enabled = config->dma_enable;
+    spi->dma_flags = config->dma_flags;
 
     lpspi_master_config_t spi_config;
     LPSPI_MasterGetDefaultConfig(&spi_config);
 
     spi_config.whichPcs = kLPSPI_Pcs0;
     spi_config.baudRate = config->baudrate;
-    spi_config.bitsPerFrame = config->datasize;
```

Comment:
I looked back at our discussion about this PR on slack, and there was definitely a confusion about this, there's no reason to remove datasize, it's definitely good to have it, and it shouldn't be a flags. The discussion was about `dmasize` not datasize.

Reply:
Uh, yeah, I mean, you wanted to save space in the struct. I literally only use the value as 8/16 so it's effectively a flag. It makes sense to make it a flag.

Reply:
I didn't want to remove data-size, SPI can use other modes than 8 and 16, also you added a duplicate of the whole config struct so no point in trying to save space anymore.

Reply:
Regarding this, it doesn't need to be in the config struct anymore. The transfer struct now controls the data size. The code will automatically update the bus/dma as required for the transfer.

For other than 8-bit/16-bit modes. I don't have anything to test with for 32-bit transfers. We don't have any device that needs that right now. Would it makes sense to add support for 32-bit transfers in the future? Or do so now.

Reply:
> This is the behavior you wanted.

No it's not, again we were talking about the DMA transfer size, not the transfer width. The transfer width does Not change between transfers, it's a set once and forget config option, it can't change between transfers without re-configuring the SPI and DMA, in fact you shouldn't be allowed to, and if it can't change why pass it as a flag in every transfer ? It's a property of the SPI/DMA config not of individual transfers.

Also if you revert the datasize in the config and add `datasize` to `omv_spi_t` there won't be any need to change the frame size every transfer, and code like this can be removed:

```
    uint32_t datasize = 8;
    if (xfer->flags & OMV_SPI_XFER_16_BIT) {
        datasize = 16;
    }

    omv_spi_set_frame_size(spi, datasize);
```